qlf_fasm/qlf_fasm.py

```diff
@@ -0,0 +1,513 @@
+#!/usr/bin/env python3
+"""
+QuickLogic qlf-series FPGA FASM to bitstream and bitstream to FASM conversion
+utility.
+"""
+import argparse
+import os
+import re
+import logging
+import pkg_resources
+
+import fasm
+
+from .database import Bit, Database
+from .bitstream import TextBitstream
+
+# =============================================================================
+
+DATABASES = {
+    "qlf_k4n8": pkg_resources.resource_filename("qlf_fasm", "database/qlf_k4n8")
+}
+
+# =============================================================================
+
+
+class QlfFasmAssembler():
+    """
+    FASM assembler for QuickLogic QLF devices.
+    """
+
+    LOC_RE = re.compile(r"(?P<name>.+)_(?P<x>[0-9]+)__(?P<y>[0-9]+)_$")
+
+    class LookupError(Exception):
+        """
+        FASM database lookup error exception
+        """
+        pass
+
+    class FeatureConflict(Exception):
+        """
+        FASM feature conflict exception
+        """
+        pass
+
+    def __init__(self, database):
+        self.bitstream = bytearray(database.bitstream_size)
```

Comment:
Shouldn't this be a Bitstream object? Maybe I missed it, but I only saw for instance `TextBitstream` instantiations, so I wonder where an actual Bitstream object may be used.

Reply:
So the intention here was to have a base class `Bitstream` and then various subclasses that implement different binary formats like `TextBitstream` and later `FourByteBitstream`. But that's not finished yet and currently only `TextBitstream` is supported.